pkg/local_object_storage/metabase/list.go

```diff
@@ -0,0 +1,208 @@
+
+	errStopIterator = errors.New("stop")
+)
+
+// ListWithCursor lists physical objects available in metabase. Includes regular,
+// tombstone and storage group objects. Does not include inhumed objects. Use
+// cursor value from response for consecutive requests.
+func ListWithCursor(db *DB, count uint32, cursor string) ([]*object.Address, string, error) {
+	r, err := db.ListWithCursor(new(ListPrm).WithCount(count).WithCursor(cursor))
+	if err != nil {
+		return nil, "", err
+	}
+
+	return r.addrList, r.cursor, nil
+}
+
+// ListWithCursor lists physical objects available in metabase. Includes regular,
+// tombstone and storage group objects. Does not include inhumed objects. Use
+// cursor value from response for consecutive requests.
+func (db *DB) ListWithCursor(prm *ListPrm) (res *ListRes, err error) {
+	err = db.boltDB.View(func(tx *bbolt.Tx) error {
+		res = new(ListRes)
+		res.addrList, res.cursor, err = db.listWithCursor(tx, prm.count, prm.cursor)
+		return err
+	})
+
+	return res, err
+}
+
+func (db *DB) listWithCursor(tx *bbolt.Tx, count int, cursor string) ([]*object.Address, string, error) {
+	threshold := len(cursor) == 0 // threshold is a flag to ignore cursor
+	a := object.NewAddress()
+	var cursorPrefix uint8
+
+	if !threshold { // if cursor is present, then decode it and check sanity
+		cursorPrefix = cursor[0]
+		switch cursorPrefix {
+		case cursorPrefixPrimary, cursorPrefixSG, cursorPrefixTombstone:
+		default:
+			return nil, "", fmt.Errorf("invalid cursor prefix %s", string(cursorPrefix))
+		}
+
+		cursor = cursor[1:]
+		if err := a.Parse(cursor); err != nil {
+			return nil, "", fmt.Errorf("invalid cursor address: %w", err)
+		}
+	}
+
+	result := make([]*object.Address, 0, count)
+	unique := make(map[string]struct{}) // do not parse the same containerID twice
+
+	_ = tx.ForEach(func(name []byte, _ *bbolt.Bucket) error {
+		containerID := parseContainerID(name, unique)
+		if containerID == nil {
+			return nil
+		}
+
+		unique[containerID.String()] = struct{}{}
+
+		if !threshold && !containerID.Equal(a.ContainerID()) {
+			return nil // ignore buckets until we find cursor bucket
+		}
+
+		prefix := containerID.String() + "/"
+
+		lookupBuckets := [...]struct {
+			name   []byte
+			prefix uint8
+		}{
+			{primaryBucketName(containerID), cursorPrefixPrimary},
+			{tombstoneBucketName(containerID), cursorPrefixTombstone},
+			{storageGroupBucketName(containerID), cursorPrefixSG},
+		}
+
+		for _, lb := range lookupBuckets {
+			if !threshold && cursorPrefix != lb.prefix {
+				continue // start from the bucket, specified in the cursor prefix
+			}
+
+			cursorPrefix = lb.prefix
+			result, cursor = selectNFromBucket(tx, lb.name, prefix, result, count, cursor, threshold)
+			if len(result) >= count {
+				return errStopIterator
+			}
+
+			// set threshold flag after first `selectNFromBucket` invocation
+			// first invocation must look for cursor object
+			threshold = true
```

Comment:
Shouldn't it be `||` ? Consider selecting from `primary` bucket, setting threshold to `true` and then iterating over the next bucket because the condition is never true.

Reply:
Not sure if understand your point.

If selecting from `primary` bucket is not enough to satisfy `if len(result) >= count` condition, then metabase goes into `tombstone` bucket with `threshold = true`, which is correct. Can you provide more details?